src/shared/sam/cli/samCliDetection.ts

```diff
@@ -37,28 +35,29 @@ export async function detectSamCli(args: { passive: boolean; showMessage: boolea
             new DefaultSamCliLocationProvider()
         )
 
-        const valueBeforeInit = samCliConfig.getSamCliLocation()
-        await samCliConfig.initialize()
-        const valueAfterInit = samCliConfig.getSamCliLocation()
-        const isAutoDetected = valueBeforeInit !== valueAfterInit
-        const isUserSettingChanged = currentsamCliLocation !== valueAfterInit
-        currentsamCliLocation = valueAfterInit
+        const sam = await samCliConfig.getOrDetectSamCli()
+        const notFound = sam.path === ''
 
-        if (args.showMessage !== false) {
-            if (!valueAfterInit) {
-                notifyUserSamCliNotDetected(samCliConfig)
-            } else if (isUserSettingChanged || args.showMessage === true) {
-                const message =
-                    !isAutoDetected && !isUserSettingChanged
-                        ? getSettingsNotUpdatedMessage(valueBeforeInit ?? '?')
-                        : getSettingsUpdatedMessage(currentsamCliLocation ?? '?')
+        // Update the user setting.
+        //
+        // NOTE: We must NOT _passively_ auto-update the user setting, that
+        // conflicts with VSCode "remote": each VSCode instance will update the
+        // setting based on its local environment, but the user settings are
+        // shared across VSCode instances...
+        if (!args.passive && sam.autoDetected) {
+            await samCliConfig.setSamCliLocation(sam.path)
+        }
```

Comment:
Instead of always calling `samCliConfig.initialize()`, only call `setSamCliLocation`, only for "active" invocation.